### PR TITLE
Maint: Fix sporadic QtDims garbage collection failures by converting some stray references to weakrefs

### PR DESCRIPTION
**Pull request by Carreau**
As far as I can tell this is due to issue with garbage collection issue
between the worker and thread. Mostly between the garbage collection and
the worker cleanup if the last reference to the thread is set to None
and collected while the thread is still active it causes a crash.

This has two side:
1) changes to make this crash more reliably, by removing unused
   reference to the thread, and using weakrefs to make sure we free up
   things faster. Freeing things faster mean we crash more reliably.
2) actually fix the cleanup to be done after thread termination and not
   worker termination. Due to 1) we are more confident this fixes
   things.

In particular, the animation thread, has a worker that refers to the
dims via a callback/slot.

We also remove the local reference to thread, to make sure there is not
a standing, unused reference to thread, thus self._animation_thread will
be the last standing reference.

That is to say we currently have:

 - worker stop
    - delete last thread reference
 (if GC, crash)
 - Thread stops.

Now we have
 - worker stop
 - Thread stops.
    - delete last thread (and worker) reference

This should fix some issues with thread/garbage collection/cycle (#5443)

### CODE REVIEW

**Comment by Carreau** [comment]
I'm not totally sure this is correct, like does the thread always stop once the worker stops ?
I think that having to deal with both the thread and worker is likely confusing, and we may want to have to deal with a single object (worker+thread), and make sure that calling stop on it, just stops everything.

**Comment by jni** [comment]
I don't fully understand the issues very well here, but (based on the conversation in #5443, nice work! I'll let @Czaki review as he is our local thread guru. 😅 Thank you both! 🙏

**Comment by jni** [comment]
Oh, btw, before merge, could we rename the PR to something more concrete, e.g., "Fix sporadic QtDims garbage collection failures by converting some stray references to weakrefs"? And try to capture some of the broader narrative in the commit message — @andy-sweet pointed out that longer commit messages (e.g. copying the whole description into the commit message) make it much nicer to investigate issues later, sometimes saving a few round trips between the terminal and the browser.

**Comment by andy-sweet** [comment]
> I'm not totally sure this is correct, like does the thread always stop once the worker stops ?

See [superqt's `new_worker_qthread`](https://github.com/pyapp-kit/superqt/blob/6ce87d44a66df9f48b61751902bebf9da0710010/src/superqt/utils/_qthreading.py#L797), which is what `AnimationWorker` uses to create the worker and thread. Based on the superqt implementation, the thread object will emit its `finished` signal after the worker emits its `finished` signal, and will do so the main event loop (because of the behavior of [`Qt::AutoConnection`](https://doc.qt.io/qt-5/qt.html#ConnectionType-enum)).

So the answer to your question is yes, but it might take a little time depending on how busy the main event loop is at the time.

File: napari/_qt/widgets/qt_dims_slider.py
Comment (andy-sweet) [inline napari/_qt/widgets/qt_dims_slider.py:494]: I don't understand why this doesn't have `qt_dims` as a parent. And I wonder if this might be an alternative fix to holding a weak reference to `QtDims`.

If I understand the QObject lifetime model correctly, I think that if the child object is still alive, then its parent must also be (i.e. the parent effectively owns the child and is controls its lifetime/destruction). So it would not be possible for `_on_click` to be executed with an invalid reference to `QtDims`.

More generally, I wonder how many of our object lifetime issues are caused by relying on Python references to Qt objects, rather than relying on the QObject hierarchy. Though that discussion is far beyond the scope of this PR :)

Comment (andy-sweet) [reply napari/_qt/widgets/qt_dims_slider.py:494]: Ah, I guess when we add the `QtPlayButton` to the layout, then reparenting will occur. So this doesn't seem like a viable alternative fix. Though I still can't explain why we need the weak reference (i.e. why would `self.qt_dims` be something we can't dereference in `self._on_click`, since it should be the grandparent of this widget).

**Comment by Carreau** [comment]
> "Fix sporadic QtDims garbage collection failures by converting some stray references to weakrefs"?

I'm not actually sure it fixes things, and I'm not totally sure I understand what is happening. But Yes. 

>  And try to capture some of the broader narrative in the commit message — @andy-sweet pointed out that longer commit messages (e.g. copying the whole description into the commit message) make it much nicer to investigate issues later, sometimes saving a few round trips between the terminal and the browser.

I've started to skip doing long messages for napari because of the squash/rebase, and relied on the PR description. So it's up to you, but the "squash" and "investigate issues later" IMHO contradict each others.

> See [superqt's new_worker_qthread](https://github.com/pyapp-kit/superqt/blob/6ce87d44a66df9f48b61751902bebf9da0710010/src/superqt/utils/_qthreading.py#L797), which is what AnimationWorker uses to create the worker and thread. Based on the superqt implementation, the thread object will emit its finished signal after the worker emits its finished signal, and will do so the main event loop (because of the behavior of [Qt::AutoConnection](https://doc.qt.io/qt-5/qt.html#ConnectionType-enum)).

> So the answer to your question is yes, but it might take a little time depending on how busy the main event loop is at the time.

Yes, that was my understanding (and how I came to this fix), but me thread two makeing are and condition race understand very well not.

File: napari/_qt/widgets/qt_dims_slider.py
Comment (Carreau) [reply napari/_qt/widgets/qt_dims_slider.py:494]: It's magic for me as well, but usually I found that weakreferences, even if not strictly necessary seem to help to have at least consistant crashes, or make it clearer when doing refactor who is the owner, and simplify reference graph when using objgraph. 

Something I haven't done there is also use wekref's callbacks to disconnect things but I don't think it's strictly necessary here.

I tend to [prefer to know reliably I'm lost](http://thecodelesscode.com/case/225).

**Comment by Carreau** [comment]
I remove the unrelated typing changes and rephrased the commit.

**Comment by jni** [comment]
> I've started to skip doing long messages for napari because of the squash/rebase, and relied on the PR description. So it's up to you, but the "squash" and "investigate issues later" IMHO contradict each others.

That's what I mean: the PR title is what by default gets put into the (final, squashed) commit message, so it should be properly descriptive.

The description *does not* get copied by default into the squashed commit but we should copy it manually on merging the PR. Sorry that I wasn't more explicit here: when I said, "capture some of the broader narrative into the commit message", I meant "the PR squash-rebase commit message", and by "we" I meant, whoever pushes the merge button. Ironically @Carreau it was one of your commits that gave rise to this discussion — it had a wonderful long description, and it could be seen in the final squashed history because it was the *only* commit in that PR. In that case, GitHub copies the full commit message into the squashed commit message.

**Comment by Carreau** [comment]
Yeah, imho this is a GH ui issue, I think that each GH PR should have a shared editable area that will reflect the final merge/squash message. Maybe in the diff view ? So that this information is also part of the Review Process.

**Comment by andy-sweet** [comment]
> The description _does not_ get copied by default into the squashed commit but we should copy it manually on merging the PR.

Looks like this might be configurable, so that the default squashed merge commit message is the PR title + description? https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests

That way the shared UI in GitHub is the PR title and description, and we don't have to rely on the merger to remember to copy the text from there (which I've been doing occasionally). They may still have to remind the author to keep the title/description up-to-date though, as here.

Unfortunately, I don't have access to napari/napari settings - I'm assuming you do @jni?

File: napari/_qt/widgets/qt_dims_slider.py
Comment (andy-sweet) [reply napari/_qt/widgets/qt_dims_slider.py:494]: I didn't read your PR description properly, but after reading this message and thinking a bit more, the weak reference is justifiable (regardless of any bugs it does or doesn't cause) because it breaks a cyclic reference between Python objects (specifically, `QtDims` and `QtPlayButton` with `QtDimSliderWidget` in the middle of the cycle). Might be good to add that as an explanatory implementation comment to the code to help the next lost soul.

In general, my guess is there's only so much pyqt/pyside can do when it comes to keeping Python and QObject lifetimes consistent/safe. Would be great if anyone knows of a resource that explains how to use those more safely (other than reading through the code itself).

Anyway, I'm pretty close to approving this. I just need to think through the dropped reference to `QtDimSliderWidget.thread`.

**Comment by jni** [comment]
> That way the shared UI in GitHub is the PR title and description, and we don't have to rely on the merger to remember to copy the text from there (which I've been doing occasionally). They may still have to remind the author to keep the title/description up-to-date though, as here.

Life changing discovery @andy-sweet!!! ❤️‍🔥 I have changed it now to default to copying title and description. 🚀 Not quite as good as @Carreau's suggestion of having the commit message also being subject to review, but much much better than what we've had until now.

@Carreau title here is still lackluster... ;)

**Comment by Carreau** [comment]
I've update the title and first message.

**Review by andy-sweet — APPROVED** [review]
Dropping `QtDimSliderWidget.thread` seems fine as the only caller of that stores that, so I'm approving this. Thanks for the investigation and fix @Carreau!

With respect to #5433, this only fixes the aborting tests, right? And not the hanging ones?

**Comment by andy-sweet** [comment]
I'll merge after 24 hours unless there are any other objections.

File: napari/_qt/widgets/qt_dims_slider.py
Comment (Czaki) [inline napari/_qt/widgets/qt_dims_slider.py]:
```suggestion
            return
```
As we do not return anything from function then plain return should be used

File: napari/_qt/widgets/qt_dims_slider.py
Comment (Czaki) [inline napari/_qt/widgets/qt_dims_slider.py]:
```suggestion
        if not qt_dims:  # pragma: no cover
```

To not generate noise in coverage reports

**Review by Czaki — APPROVED** [review]
Small style remarks. I think that I understand how it works, but I'm not 100% sure

**Comment by Carreau** [comment]
> I think that I understand how it works, but I'm not 100% sure

We are in the same boat.

> With respect to #5433, this only fixes the aborting tests, right? And not the hanging ones?

It's one problem with flaky test you can never be sure. 

One thing I can suggest is to set the 

`job.<id>.<steps?>.timeout-minutes` for each GH action item to at least get a quick failure.
Say current job time + 10% ? Possibly only the tox step to be conservative with what can hang ?

**Comment by Czaki** [comment]
> Say current job time + 10% ? Possibly only the tox step to be conservative with what can hang ?

My observation is that sometimes, because of network/pypi problems 10 minutes buffer is required, but we still could use 40 minutes cap instead of six hours.

**Comment by Carreau** [comment]
Opened task to do so as #5488

**Comment by andy-sweet** [comment]
> > With respect to #5433, this only fixes the aborting tests, right? And not the hanging ones?
> 
> It's one problem with flaky test you can never be sure.

Sure. But this PR offers one explanation of the aborting test (by attempting to dereference a C++ object that has been deleted) and proposes a fix for it. By contrast, I don't see how it could explain a hanging test - would be happy if it does though, especially with an explanation!

**Pull request merged**
